x/merkledb/codec.go

```diff
@@ -107,30 +103,54 @@ func encodeDBNode(n *dbNode) []byte {
 	return buf.Bytes()
 }
 
-// Returns the bytes that will be hashed to generate [n]'s ID.
+// Returns the canonical hash of [n].
+//
 // Assumes [n] is non-nil.
-func encodeHashValues(n *node) []byte {
+// This method is performance critical. It is not expected to perform any memory
+// alloactions.
+func hashNode(n *node) ids.ID {
 	var (
-		numChildren = len(n.children)
-		// Estimate size [hv] to prevent memory allocations
-		estimatedLen = minVarIntLen + numChildren*hashValuesChildLen + estimatedValueLen + estimatedKeyLen
-		buf          = bytes.NewBuffer(make([]byte, 0, estimatedLen))
+		sha  = sha256.New()
+		hash ids.ID
+		// The hash length is larger than the maximum Uvarint length. This
+		// ensures binary.AppendUvarint doesn't perform any memory allocations.
+		emptyHashBuffer = hash[:0]
 	)
 
-	encodeUint(buf, uint64(numChildren))
+	// By directly calling sha.Write rather than passing sha around as an
+	// io.Writer, the compiler can perform sufficient escape analysis to avoid
+	// allocating buffers on the heap.
+	_, _ = sha.Write(binary.AppendUvarint(emptyHashBuffer, uint64(len(n.children))))
 
-	// ensure that the order of entries is consistent
-	keys := maps.Keys(n.children)
+	// By allocating BranchFactorLargest rather than len(n.children), this slice
+	// is allocated on the stack rather than the heap. BranchFactorLargest is
+	// at least len(n.children) which avoids memory allocations.
+	keys := make([]byte, 0, BranchFactorLargest)
+	for k := range n.children {
```

Comment:
dumb question: is there a reason we don't just make this an array where the key is the index?

Reply:
> is there a reason we don't just make this an array where the key is the index?

None that I can think of.

Reply:
If we did it to save memory, it seems like these extra ops we need to do during encoding are probably not worth it (not to mention that it may not actually save).